File: A/415-add-strings/main.go
package main

import "fmt"

func addStrings(num1 string, num2 string) string {
	l := len(num1)
	if len(num1) < len(num2) {
		l = len(num2)
	}
	buf := make([]byte, l+1)
	for i := 0; i < len(buf); i++ {
		buf[i] = ' '
	}
	i1, i2 := 0, 0
	isPlus := false
	for i1 < len(num1) && i2 < len(num2) {
		c1 := int(num1[len(num1)-1-i1] - '0')
		c2 := int(num2[len(num2)-1-i2] - '0')
		c := c1 + c2
		if isPlus {
			c++
		}
		if c > 9 {
			isPlus = true
			c -= 10
		} else {
			isPlus = false
		}
		buf[l-i1] = byte(int('0') + c)
		i1++
		i2++
	}
	if i1 == len(num1) {
		for i := i1; i < len(num2); i++ {
			c := int(num2[len(num2)-1-i] - '0')
			if isPlus {
				c++
			}
			if c > 9 {
				isPlus = true
				c -= 10
			} else {
				isPlus = false
			}
			buf[l-i] = byte(int('0') + c)
		}
	}
	if i2 == len(num2) {
		for i := i1; i < len(num1); i++ {
			c := int(num1[len(num1)-1-i] - '0')
			if isPlus {
				c++
			}
			if c > 9 {
				isPlus = true
				c -= 10
			} else {
				isPlus = false
			}
			buf[l-i] = byte(int('0') + c)
		}
	}
	if isPlus {
		buf[0] = '1'
	}

	if buf[0] == ' ' {
		buf = buf[1:]
	}
	return string(buf)
}

func main() {
	fmt.Println("a")
}

File: A/415-add-strings/main.rs
struct Solution {}

impl Solution {
  pub fn add_strings(num1: String, num2: String) -> String {
    let mut idx1: i32 = num1.len() as i32 - 1;
    let mut idx2: i32 = num2.len() as i32 - 1;
    let b1 = num1.as_bytes();
    let b2 = num2.as_bytes();
    let mut is_carry = false;
    let mut buf: Vec<u8> = Vec::new();
    while idx1 >= 0 && idx2 >= 0 {
      let n1 = (b1[idx1 as usize] - '0' as u8) as i32;
      let n2 = (b2[idx2 as usize] - '0' as u8) as i32;
      let mut sum = n1 + n2;
      if is_carry {
        sum += 1;
      }
      if sum > 9 {
        sum -= 10;
        is_carry = true;
      } else {
        is_carry = false;
      }
      buf.push((sum as u8 + '0' as u8) as u8);

      idx1 -= 1;
      idx2 -= 1;
    }

    if idx1 >= 0 {
      while idx1 >= 0 {
        let n = (b1[idx1 as usize] - '0' as u8) as i32;
        let mut sum = n;
        if is_carry {
          sum += 1;
        }
        if sum > 9 {
          sum -= 10;
          is_carry = true;
        } else {
          is_carry = false;
        }
        buf.push((sum as u8 + '0' as u8) as u8);

        idx1 -= 1;
      }
    }

    if idx2 >= 0 {
      while idx2 >= 0 {
        let n = (b2[idx2 as usize] - '0' as u8) as i32;
        let mut sum = n;
        if is_carry {
          sum += 1;
        }
        if sum > 9 {
          sum -= 10;
          is_carry = true;
        } else {
          is_carry = false;
        }
        buf.push((sum as u8 + '0' as u8) as u8);
        idx2 -= 1;
      }
    }

    if is_carry {
      buf.push('1' as u8);
    }
    let mut start: usize = 0;
    let mut end: usize = buf.len() - 1;
    while start < end {
      let tmp = buf[start];
      buf[start] = buf[end];
      buf[end] = tmp;
      start += 1;
      end -= 1;
    }
    String::from_utf8(buf).unwrap()
  }
}
